### PR TITLE
Laboratory work #3, Dasha Oleinik - 23FPL1

### PR DESCRIPTION
**Pull request by dkole13**

### CODE REVIEW

File: lab_3_ann_retriever/main.py [hunk 1]
```diff
@@ -49,6 +52,13 @@ def calculate_distance(query_vector: Vector, document_vector: Vector) -> float |
 
     In case of corrupt input arguments, None is returned.
     """
+    if query_vector is None or document_vector is None:
+        return None
+
+    if not query_vector or not document_vector:
+        return 0.0
+
+    return sqrt(sum((query_vector[i] - document_vector[i]) ** 2 for i in range(len(query_vector))))
```
Comment (EIzharikov) [inline lab_3_ann_retriever/main.py:61]: try using ```zip```

File: lab_3_ann_retriever/main.py [hunk 2]
```diff
@@ -157,6 +195,32 @@ def build(self) -> bool:
...
+                    raise ValueError
+                if token not in self._vocabulary:
+                    self._vocabulary.append(token)
+        self._vocabulary.sort()
+
+        if self._vocabulary is None:
+            raise ValueError
+
+        self._token2ind = {}
+        for i in range(len(self._vocabulary)):
```
Comment (EIzharikov) [inline lab_3_ann_retriever/main.py:214]: try ```enumerate``` here and list comprehension

File: lab_3_ann_retriever/main.py [hunk 3]
```diff
@@ -220,6 +299,14 @@ def _calculate_tf_idf(self, document: list[str]) -> Vector | None:
 
         In case of corrupt input arguments, None is returned.
         """
+        if not isinstance(document, list) or not all(isinstance(token, str) for token in document) \
+                or not document:
+            return None
+
+        self.build()
```
Comment (EIzharikov) [inline lab_3_ann_retriever/main.py:306]: why are you bulding in calculate??

File: lab_3_ann_retriever/main.py [hunk 4]
```diff
@@ -269,6 +367,22 @@ def retrieve_relevant_documents(
...
+
+        query_vectorized = self._index_document(query)
+        if query_vectorized is None:
+            return None
+        self.index_documents(self._documents)
+        knn = self._calculate_knn(query_vectorized, self._document_vectors, n_neighbours)
+        if knn is None or not knn:
+            return None
+        relevant_docs = []
+        for pair in knn:
```
Comment (EIzharikov) [inline lab_3_ann_retriever/main.py:382]: divide into index and value here

File: lab_3_ann_retriever/main.py [hunk 5]
```diff
@@ -321,6 +437,15 @@ def _calculate_knn(
 
         In case of corrupt input arguments, None is returned.
         """
+        if not isinstance(document_vectors, list) or not isinstance(n_neighbours, int) or \
+                not query_vector:
+            return None
+
+        distances = []
+        for i in range(len(document_vectors)):
```
Comment (EIzharikov) [inline lab_3_ann_retriever/main.py:445]: enumerate!!

File: lab_3_ann_retriever/main.py [hunk 6]
```diff
@@ -321,6 +437,15 @@ def _calculate_knn(
...
+            distances.append((i, calculate_distance(query_vector, document_vectors[i])))
+        distances = sorted(distances, key=lambda t: t[1])
```
Comment (EIzharikov) [inline lab_3_ann_retriever/main.py:447]: what is t?

**Review by EIzharikov — COMMENTED** [review]
Nice progress! Make CI green and answer&fix all my comments. Good luck

File: lab_3_ann_retriever/main.py (hunk 1 above)
Comment (dkole13) [reply lab_3_ann_retriever/main.py:61]: done

File: lab_3_ann_retriever/main.py (hunk 2 above)
Comment (dkole13) [reply lab_3_ann_retriever/main.py:214]: done

File: lab_3_ann_retriever/main.py (hunk 4 above)
Comment (dkole13) [reply lab_3_ann_retriever/main.py:382]: done

File: lab_3_ann_retriever/main.py (hunk 5 above)
Comment (dkole13) [reply lab_3_ann_retriever/main.py:445]: done

File: lab_3_ann_retriever/main.py (hunk 6 above)
Comment (dkole13) [reply lab_3_ann_retriever/main.py:447]: renamed

File: lab_3_ann_retriever/main.py (hunk 3 above)
Comment (EIzharikov) [reply lab_3_ann_retriever/main.py:295]: so, may be tests are irrelevant or you have mistake in another place. Check it, please

**Review by EIzharikov — COMMENTED** [review]
Nice progress! You have almost done with KDTree! It is awesome! Make CI green, fix&answer all my comments and continue your hard work!

File: lab_3_ann_retriever/main.py [hunk 7]
```diff
@@ -321,6 +433,18 @@ def _calculate_knn(
 
         In case of corrupt input arguments, None is returned.
         """
+        if not isinstance(document_vectors, list) or not isinstance(n_neighbours, int) or \
+                not query_vector or not document_vectors:
+            return None
+
+        distances: list[tuple[int, float]] = []
```
[the remark this answers is not in the document]
Comment (dkole13) [reply lab_3_ann_retriever/main.py:440]: i did that in order to remove a mypy error, but now it seems to be working without the typing. deleted

File: lab_3_ann_retriever/main.py [hunk 8]
```diff
@@ -426,6 +565,54 @@ def build(self, vectors: list[Vector]) -> bool:
...
+
+        states_info = [{
+                           'vectors': [(vector, index) for index, vector in enumerate(vectors)],
+                           'depth': 0,
+                           'parent': Node(tuple([0.0] * len(vectors[0])), -1),
+                           'is_left': True
+                       }]
+
+        while states_info:
+            current_vectors: list[tuple[tuple[float], int]] = states_info[0]['vectors']
```
[the remark this answers is not in the document]
Comment (dkole13) [reply lab_3_ann_retriever/main.py:579]: without it i get 5 additional mypy errors :(

File: lab_3_ann_retriever/main.py [hunk 9]
```diff
@@ -426,6 +565,54 @@ def build(self, vectors: list[Vector]) -> bool:
...
+            is_left = states_info[0]['is_left']
+            states_info.pop(0)
+            if current_vectors:
+                axis = depth % len(current_vectors[0])
+                current_vectors.sort(key=lambda vector: vector[0][axis])
+                median_index = len(current_vectors) // 2
+                median_node = Node(current_vectors[median_index][0],
+                                   current_vectors[median_index][1])
+
+                if parent.payload == -1:
```
[the remark this answers is not in the document]
Comment (dkole13) [reply lab_3_ann_retriever/main.py:591]: done

File: lab_3_ann_retriever/main.py (hunk 3 above)
Comment (dkole13) [reply lab_3_ann_retriever/main.py:295]: finally managed to fix it!

File: lab_3_ann_retriever/main.py [hunk 10]
```diff
@@ -117,6 +154,20 @@ def tokenize_documents(self, documents: list[str]) -> list[list[str]] | None:
 
         In case of corrupt input arguments, None is returned.
         """
+        if not isinstance(documents, list) or not all(isinstance(doc, str) for doc in documents) \
+                or not documents:
+            return None
+
+        tokenized_documents: list[list[str]] = []
```
Comment (EIzharikov) [inline lab_3_ann_retriever/main.py:161]: remove direct typing

File: lab_3_ann_retriever/main.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (EIzharikov) [reply lab_3_ann_retriever/main.py:699]: yes, fix it wit another way, not straight typings, its too easy

[the remark this answers is not in the document]
Comment (EIzharikov) [reply lab_3_ann_retriever/main.py:699]: and unfair

File: lab_3_ann_retriever/start.py [hunk 11]
```diff
@@ -28,9 +30,31 @@ def main() -> None:
...
+
+    documents, stopwords = open_files()
+
+    tokenizer = Tokenizer(stopwords)
+    tokenized_documents = tokenizer.tokenize_documents(documents)
+
+    vectorizer = Vectorizer(tokenized_documents)
+    vectorizer.build()
+
+    # secret_vector = tuple(float(value.replace(',', '')) for value in text.split())
```
[the remark this answers is not in the document]
Comment (EIzharikov) [reply lab_3_ann_retriever/start.py:45]: delete of course

**Review by EIzharikov — COMMENTED** [review]
nice progress!! Make CI green and be ready for the friday!

File: lab_3_ann_retriever/main.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (dkole13) [reply lab_3_ann_retriever/main.py]: removed

File: lab_3_ann_retriever/start.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (dkole13) [reply lab_3_ann_retriever/start.py]: done

File: lab_3_ann_retriever/main.py (hunk 3 above)
Comment (EIzharikov) [reply lab_3_ann_retriever/main.py]: NICE!!!

**Review by EIzharikov — COMMENTED** [review]
Great progress! Fix&answer comments and be ready for the Friday!

File: lab_3_ann_retriever/main.py [hunk 12]
```diff
@@ -220,13 +326,21 @@ def _calculate_tf_idf(self, document: list[str]) -> Vector | None:
...
+            tf[word] = document.count(word) / len(document)
+        return tuple((tf[word] * self._idf_values[word] if word in document else 0.0
+                      for word in self._vocabulary))
 
 
 class BasicSearchEngine:
     """
     Engine based on KNN algorithm.
     """
-
```
[the remark this answers is not in the document]
Comment (dkole13) [reply lab_3_ann_retriever/main.py]: done

File: lab_3_ann_retriever/main.py [hunk 13]
```diff
@@ -353,13 +548,28 @@ def _load_documents(self, state: dict) -> bool:
...
+                return False
+            self._document_vectors.append(loaded_vector)
+        return True
 
 
 class Node(NodeLike):
     """
     Interface definition for Node for KDTree.
     """
-
```
[the remark this answers is not in the document]
Comment (dkole13) [reply lab_3_ann_retriever/main.py]: done

File: lab_3_ann_retriever/main.py [hunk 14]
```diff
@@ -426,6 +676,50 @@ def build(self, vectors: list[Vector]) -> bool:
 
         In case of corrupt input arguments, False is returned.
         """
+        if not isinstance(vectors, list) or not vectors:
+            return False
+
+        states_info: list[dict] = [{
```
[the remark this answers is not in the document]
Comment (dkole13) [reply lab_3_ann_retriever/main.py]: done

File: lab_3_ann_retriever/main.py [hunk 15]
```diff
@@ -426,6 +676,50 @@ def build(self, vectors: list[Vector]) -> bool:
...
+                           'depth': 0,
+                           'parent': Node(tuple([0.0] * len(vectors[0])), -1),
+                           'is_left': True
+                       }]
+        while states_info:
+            current_vectors = states_info[0]['vectors']
+            depth = states_info[0]['depth']
+            parent = states_info[0]['parent']
+            is_left = states_info[0]['is_left']
+            states_info.pop(0)
```
[the remark this answers is not in the document]
Comment (dkole13) [reply lab_3_ann_retriever/main.py]: done

File: lab_3_ann_retriever/main.py [hunk 16]
```diff
@@ -495,13 +823,42 @@ def _find_closest(self, vector: Vector, k: int = 1) -> list[tuple[float, int]] |
...
+                elif node.right_node is not None:
+                    nodes.append((node.left_node, new_depth))
+        return nearest_nodes
 
 
 class SearchEngine(BasicSearchEngine):
     """
     Retriever based on KDTree algorithm.
     """
-
```
[the remark this answers is not in the document]
Comment (dkole13) [reply lab_3_ann_retriever/main.py]: done

**Pull request closed without merging**